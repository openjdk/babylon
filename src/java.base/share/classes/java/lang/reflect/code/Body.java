/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.lang.reflect.code;

import java.lang.reflect.code.type.FunctionType;
import java.util.*;

/**
 * A body containing a sequence of (basic) blocks.
 * <p>
 * The sequence of blocks form a graph. The last operation in a block, a terminating operation,
 * may refer to other blocks in the sequence as successors, thus forming the graph. Otherwise, the last
 * operation defines how the body passes control flow back to the parent operation, and in doing so may optionally
 * yield a value.
 * <p>
 * The first block in the sequence is the entry block, and no other blocks refer to it as a successor.
 * <p>
 * A body has a function type whose return type is the body's yield type and whose parameter types are the entry
 * block's parameters types, in order.
 * The function type describes the sequence of input parameters types for arguments that are passed to the
 * body when control flow is passed to it, and describes the return type of values that are returned when body passes
 * control back to its parent operation.
 */
public final class Body implements CodeElement<Body, Block> {
    // Parent operation
    // Non-null when body is built, and therefore bound to operation
    Op parentOp;

    // The ancestor body, when null the body is isolated and cannot refer to values defined outside
    // When non-null and body is built, ancestorBody == parentOp.result.block.parentBody
    final Body ancestorBody;

    final TypeElement yieldType;

    // Sorted in reverse postorder
    final List<Block> blocks;

    // Map of a block to its immediate dominator
    // Computed lazily, null if not computed
    Map<Block, Block> idoms;

    /**
     * Constructs a body, whose ancestor is the given ancestor body.
     */
    Body(Body ancestorBody, TypeElement yieldType) {
        this.ancestorBody = ancestorBody;
        this.yieldType = yieldType;
        this.blocks = new ArrayList<>();
    }

    /**
     * Returns this body's parent operation.
     *
     * @return the body's parent operation.
     */
    @Override
    public Op parent() {
        return parentOp;
    }

    /**
     * Returns this body's parent operation.
     *
     * @return the body's parent operation.
     */
    public Op parentOp() {
        return parentOp;
    }

    @Override
    public List<Block> children() {
        return blocks();
    }

    /**
     * Returns body's blocks in reverse-postorder as an unmodifiable list.
     *
     * @return the body's blocks in reverse-postorder.
     */
    public List<Block> blocks() {
        return Collections.unmodifiableList(blocks);
    }

    /**
     * {@return the yield type of this body}
     */
    public TypeElement yieldType() {
        return yieldType;
    }

    /**
     * Returns the body's function type.
     * <p>The function type is composed of the body's entry block parameter types and
     * the body's yield type.
     *
     * @return the body type.
     */
    public FunctionType bodyType() {
        Block entryBlock = entryBlock();
        return FunctionType.functionType(yieldType, entryBlock.parameterTypes());
    }

    /**
     * Finds the block in this body that is the ancestor of the given block.
     *
     * @param b the given block.
     * @return the block in this body that is the ancestor of the given block,
     * otherwise {@code null}
     */
    public Block findAncestorBlockInBody(Block b) {
        Objects.requireNonNull(b);

        while (b != null && b.parentBody() != this) {
            b = b.parentBody().parentOp().parentBlock();
        }

        return b;
    }

    /**
     * Returns this body's entry block.
     * <p>
     * The entry block is the first block in the sequence. No other blocks refer to it as a successor.
     *
     * @return the body's entry block
     */
    public Block entryBlock() {
        return blocks.getFirst();
    }

    /**
     * Returns a map of block to its immediate dominator.
     *
     * @return a map of block to its immediate dominator
     */
    public Map<Block, Block> immediateDominators() {
        /*
         * Compute dominators of blocks in a body.
         * <p>
         * https://www.cs.rice.edu/~keith/EMBED/dom.pdf
         * A Simple, Fast Dominance Algorithm
         * Keith D. Cooper, Timothy J. Harvey, and Ken Kennedy
         */

        if (idoms != null) {
            return idoms;
        }

        Map<Block, Block> doms = idoms = new HashMap<>();
        doms.put(entryBlock(), entryBlock());

        // Blocks are sorted in reverse postorder
        boolean changed;
        do {
            changed = false;
            // Iterate through blocks in reverse postorder, except for entry block
            for (int i = 1; i < blocks.size(); i++) {
                Block b = blocks.get(i);

                // Find first processed predecessor of b
                Block newIdom = null;
                for (Block p : b.predecessors()) {
                    if (doms.containsKey(p)) {
                        newIdom = p;
                        break;
                    }
                }
                assert b.predecessors().isEmpty() || newIdom != null : b;

                // For all other predecessors, p, of b
                for (Block p : b.predecessors()) {
                    if (p == newIdom) {
                        continue;
                    }

                    if (doms.containsKey(p)) {
                        // If already calculated
                        newIdom = intersect(doms, p, newIdom);
                    }
                }

                if (doms.get(b) != newIdom) {
                    doms.put(b, newIdom);
                    changed = true;
                }
            }
        } while (changed);

        return doms;
    }

    static Block intersect(Map<Block, Block> doms, Block b1, Block b2) {
        while (b1 != b2) {
            while (b1.index > b2.index) {
                b1 = doms.get(b1);
            }

            while (b2.index > b1.index) {
                b2 = doms.get(b2);
            }
        }

        return b1;
    }

    /**
     * Returns the dominance frontier of each block in the body.
     * <p>
     * The dominance frontier of block, {@code B} say, is the set of all blocks, {@code C} say,
     * such that {@code B} dominates a predecessor of {@code C} but does not strictly dominate
     * {@code C}.
     *
     * @return the dominance frontier of each block in the body
     */
    public Map<Block, Set<Block>> dominanceFrontier() {
        // @@@ cache result?
        Map<Block, Block> idoms = immediateDominators();
        Map<Block, Set<Block>> df = new HashMap<>();

        for (Block b : blocks) {
            Set<Block> preds = b.predecessors();

            if (preds.size() > 1) {
                for (Block p : preds) {
                    Block runner = p;
                    while (runner != idoms.get(b)) {
                        df.computeIfAbsent(runner, k -> new LinkedHashSet<>()).add(b);
                        runner = idoms.get(runner);
                    }
                }
            }
        }

        return df;
    }

    /**
     * Returns {@code true} if this body is dominated by the given body {@code dom}.
     * <p>
     * A body, {@code b} say, is dominated by {@code dom} if {@code b} is the same as {@code dom} or a descendant of
     * {@code dom}. Specifically, if {@code b} and {@code dom} are not equal then {@code b} becomes the nearest ancestor
     * body, the result of {@code b.parentOp().parentBlock().parentBody()}, and so on until either:
     * {@code b == dom}, therefore {@code b} is dominated by {@code dom} and this method returns {@code true};
     * or {@code b.parentOp().parentBlock() == null}, therefore {@code b} is <b>not</b> dominated
     * by {@code dom} and this method returns {@code false}.
     *
     * @param dom the dominating body
     * @return {@code true} if this body is dominated by the given body {@code dom}.
     */
    public boolean isDominatedBy(Body dom) {
        return isDominatedBy(this, dom);
    }

    static boolean isDominatedBy(Body r, Body dom) {
        while (r != dom) {
            Block eb = r.parentOp().parentBlock();
            if (eb == null) {
                return false;
            }

            r = eb.parentBody();
        }

        return true;
    }

    /**
     * A builder of a body.
     * <p>
     * When the body builder is built any associated block builders are also considered built.
     */
    public final class Builder {
        /**
         * Creates a body build with a new context, and a copying transformer.
         *
         * @param ancestorBody the nearest ancestor body builder
         * @param bodyType     the body's function type
         * @return the body builder
         * @throws IllegalStateException if the ancestor body builder is built
         * @see #of(Builder, FunctionType, CopyContext, OpTransformer)
         */
        public static Builder of(Builder ancestorBody, FunctionType bodyType) {
            // @@@ Creation of CopyContext
            return of(ancestorBody, bodyType, CopyContext.create(), OpTransformer.COPYING_TRANSFORMER);
        }

        /**
         * Creates a body build with a copying transformer.
         *
         * @param ancestorBody the nearest ancestor body builder
         * @param bodyType     the body's function type
         * @param cc           the context
         * @return the body builder
         * @throws IllegalStateException if the ancestor body builder is built
         * @see #of(Builder, FunctionType, CopyContext, OpTransformer)
         */
        public static Builder of(Builder ancestorBody, FunctionType bodyType, CopyContext cc) {
            return of(ancestorBody, bodyType, cc, OpTransformer.COPYING_TRANSFORMER);
        }

        /**
         * Creates a body builder.
         * <p>
         * Structurally, the created body builder must be built before its ancestor body builder (if non-null) is built,
         * otherwise an {@code IllegalStateException} will occur.
         * <p>
         * A function type, referred to as the body type, defines the body's yield type and the initial sequence of
         * entry block parameters.
         * The body's yield type is the function type's return type.
         * An entry block builder is created with appended block parameters corresponding, in order, to
         * the function type's parameter types.
         * <p>
         * If the ancestor body is null then the created body builder is isolated and descendant operations may only
         * refer to values declared within the created body builder. Otherwise, operations
         * may refer to values declared in the ancestor body builders (outside the created body builder).
         *
         * @param ancestorBody the nearest ancestor body builder, may be null if isolated
         * @param bodyType     the body's function type
         * @param cc           the context
         * @param ot           the transformer
         * @return the body builder
         * @throws IllegalStateException if the ancestor body builder is built
         * @see #of(Builder, FunctionType, CopyContext, OpTransformer)
         */
        public static Builder of(Builder ancestorBody, FunctionType bodyType,
                                 CopyContext cc, OpTransformer ot) {
            Body body = new Body(ancestorBody != null ? ancestorBody.target() : null, bodyType.returnType());
            return body.new Builder(ancestorBody, bodyType, cc, ot);
        }

        // The ancestor body, may be null
        final Builder ancestorBody;

        // The entry block of this body, whose parameters are given by the body's function type
        final Block.Builder entryBlock;

        // When non-null contains one or more great-grandchildren
        List<Builder> greatgrandchildren;

        // True when built
        boolean closed;

        Builder(Builder ancestorBody, FunctionType bodyType,
                CopyContext cc, OpTransformer ot) {
            // Structural check
            // The ancestor body should not be built before this body is created
            if (ancestorBody != null) {
                ancestorBody.check();
                ancestorBody.addGreatgrandchild(this);
            }

            this.ancestorBody = ancestorBody;
            // Create entry block from the body's function type
            Block eb = Body.this.createBlock(bodyType.parameterTypes());
            this.entryBlock = eb.new Builder(this, cc, ot);
        }

        void addGreatgrandchild(Builder greatgrandchild) {
            var l = greatgrandchildren == null
                    ? (greatgrandchildren = new ArrayList<>()) : greatgrandchildren;
            l.add(greatgrandchild);
        }

        /**
         * Builds the body and its blocks, associating the body with a parent operation.
         * <p>
         * Structurally, any descendant body builders must be built before this body builder is built,
         * otherwise an {@code IllegalStateException} will occur.
         * <p>
         * Blocks are sorted in reserve postorder.
         * <p>
         * Any unreferenced empty blocks are removed. An unreferenced block is a non-entry block with no predecessors.
         *
         * @param op the parent operation
         * @return the build body
         * @throws IllegalStateException if this body builder is built
         * @throws IllegalStateException if any descendant body builders are not built
         * @throws IllegalStateException if a block has no terminal operation, unless unreferenced and empty
         */
        // @@@ Validation
        // e.g., every operand dominates the operation result (potentially expensive)
        public Body build(Op op) {
            // Structural check
            // This body should not be closed
            check();
            closed = true;

            // Structural check
            // All great-grandchildren bodies should be built
            if (greatgrandchildren != null) {
                for (Builder greatgrandchild : greatgrandchildren) {
                    if (!greatgrandchild.closed) {
                        throw new IllegalStateException("Descendant body builder is not built");
                    }
                }
            }

            Iterator<Block> i = blocks.iterator();
            while (i.hasNext()) {
                Block block = i.next();

                // Structural check
                // All referenced blocks should have a terminating operation as the last operation
                if (block.ops.isEmpty()) {
                    if (block.isEntryBlock() || !block.predecessors.isEmpty()) {
                        throw noTerminatingOperation();
                    }

                    // Remove unreferenced empty block
                    assert !block.isEntryBlock() && block.predecessors.isEmpty();
                    i.remove();
                } else if (!(block.ops.getLast() instanceof Op.Terminating)) {
                    throw noTerminatingOperation();
                }
            }

            sortReversePostorder();

            Body.this.parentOp = op;
            return Body.this;
        }

        static IllegalStateException noTerminatingOperation() {
            return new IllegalStateException("Block has no terminating operation as the last operation");
        }

        /**
         * Returns the body builder's function type.
         * <p>
         * The function type is composed of the body builder's yield type, as the function type's return type, and the
         * currently built entry block's parameter types, in order, as the function type's parameter types.
         *
         * @return the body builder's function type
         */
        public FunctionType bodyType() {
            TypeElement returnType = Body.this.yieldType();
            Block eb = Body.this.entryBlock();
            return FunctionType.functionType(returnType, eb.parameterTypes());
        }

        /**
         * {@return the body builder's nearest ancestor body builder}
         */
        public Builder ancestorBody() {
            return ancestorBody;
        }

        /**
         * {@return the body's entry block builder}
         */
        public Block.Builder entryBlock() {
            return entryBlock;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            return o instanceof Builder that && Body.this == that.target();
        }

        @Override
        public int hashCode() {
            return Body.this.hashCode();
        }

        void check() {
            if (closed) {
                throw new IllegalStateException("Builder is closed");
            }
        }

        Body target() {
            return Body.this;
        }

        // Build new block in body
        Block.Builder block(List<TypeElement> params, CopyContext cc, OpTransformer ot) {
            check();
            Block block = Body.this.createBlock(params);

            return block.new Builder(this, cc, ot);
        }
    }

    /**
     * Copies the contents of this body.
     *
     * @param cc the copy context
     * @return the builder of a body containing the copied body
     * @see #transform(CopyContext, OpTransformer)
     */
    public Builder copy(CopyContext cc) {
        return transform(cc, OpTransformer.COPYING_TRANSFORMER);
    }

    /**
     * Transforms this body.
     * <p>
     * A new body builder is created with the same function type as this body.
     * Then, this body is {@link Block.Builder#transformBody(Body, java.util.List, CopyContext, OpTransformer) transformed}
     * into the body builder's entry block builder with the given copy context, operation transformer, and values
     * that are the entry block's parameters.
     *
     * @param cc the copy context
     * @param ot the operation transformer
     * @return a body builder containing the transformed body
     */
    public Builder transform(CopyContext cc, OpTransformer ot) {
        Block.Builder ancestorBlockBuilder = ancestorBody != null
                ? cc.getBlock(ancestorBody.entryBlock()) : null;
        Builder ancestorBodyBuilder = ancestorBlockBuilder != null
                ? ancestorBlockBuilder.parentBody() : null;
        Builder body = Builder.of(ancestorBodyBuilder,
                // Create function type with just the return type and add parameters afterward
                FunctionType.functionType(yieldType),
                cc, ot);

        for (Block.Parameter p : entryBlock().parameters()) {
            body.entryBlock.parameter(p.type());
        }

        body.entryBlock.transformBody(this, body.entryBlock.parameters(), cc, ot);
        return body;
    }

    // Sort blocks in reverse post order
    // After sorting the following holds for a block
    //   block.parentBody().blocks().indexOf(block) == block.index()
    private void sortReversePostorder() {
        if (blocks.size() < 2) {
            for (int i = 0; i < blocks.size(); i++) {
                blocks.get(i).index = i;
            }
            return;
        }

        // Reset block indexes
        // Also ensuring blocks with no predecessors occur last
        for (Block b : blocks) {
            b.index = Integer.MAX_VALUE;
        }

        Deque<Block> stack = new ArrayDeque<>();
        stack.push(blocks.get(0));

        // Postorder iteration
        int index = blocks.size();
        while (!stack.isEmpty()) {
            Block n = stack.peek();
            if (n.index == Integer.MIN_VALUE) {
                // If n's successor has been processed then add n
                stack.pop();
                n.index = --index;
            } else if (n.index < Integer.MAX_VALUE) {
                // If n has already been processed then ignore
                stack.pop();
            } else {
                // Mark before processing successors, a successor may refer back to n
                n.index = Integer.MIN_VALUE;
                for (Block.Reference s : n.successors()) {
                    if (s.target.index < Integer.MAX_VALUE) {
                        continue;
                    }

                    stack.push(s.target);
                }
            }
        }

        blocks.sort(Comparator.comparingInt(b -> b.index));
        if (blocks.get(0).index > 0) {
            // There are blocks with no predecessors
            // Reassign indexes to their natural indexes, sort order is preserved
            for (int i = 0; i < blocks.size(); i++) {
                blocks.get(i).index = i;
            }
        }
    }

    // Modifying methods

    // Create block
    private Block createBlock(List<TypeElement> params) {
        Block b = new Block(this, params);
        blocks.add(b);
        return b;
    }
}
