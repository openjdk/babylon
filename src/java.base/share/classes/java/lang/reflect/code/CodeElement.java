/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.lang.reflect.code;

import java.util.List;
import java.util.function.BiFunction;
import java.util.function.Predicate;
import java.util.stream.Stream;

/**
 * A code element, one of {@link Body body}, {@link Block block}, or {@link Op operation}.
 * <p>
 * A code element may have child code elements, and so on, to form a tree. A (root) code element and all its descendants
 * can be traversed.
 *
 * @param <E> the code element type
 * @param <C> the child code element type.
 */
// @@@ E may not be needed
public sealed interface CodeElement<
        E extends CodeElement<E, C>,
        C extends CodeElement<C, ?>>
        extends CodeItem
        permits Body, Block, Op {

    /**
     * {@return a stream of code elements sorted topologically in pre-order traversal.}
     */
    // Code copied into the compiler cannot depend on new gatherer API
    default Stream<CodeElement<?, ?>> elements() {
/*__throw new UnsupportedOperationException();__*/        return Stream.of(Void.class).gather(() -> (_, _, downstream) -> traversePreOrder(downstream::push));
    }

//    private boolean traversePreOrder(Gatherer.Downstream<? super CodeElement<?, ?>> v) {
    private boolean traversePreOrder(Predicate<? super CodeElement<?, ?>> v) {
        if (!v.test(this)) {
            return false;
        }
        for (C c : children()) {
            if (!((CodeElement<?, ?>) c).traversePreOrder(v)) {
                return false;
            }
        }
        return true;
    }

    /**
     * Traverses this code element and any descendant code elements.
     * <p>
     * Traversal is performed in pre-order, reporting each code element to the visitor.
     *
     * @param t   the traversing accumulator
     * @param v   the code element visitor
     * @param <T> accumulator type
     * @return the traversing accumulator
     */
    default <T> T traverse(T t, BiFunction<T, CodeElement<?, ?>, T> v) {
        t = v.apply(t, this);
        for (C r : children()) {
            t = r.traverse(t, v);
        }

        return t;
    }

    /**
     * Creates a visiting function for bodies.
     *
     * @param v   the body visitor
     * @param <T> accumulator type
     * @return the visiting function for bodies
     */
    static <T> BiFunction<T, CodeElement<?, ?>, T> bodyVisitor(BiFunction<T, Body, T> v) {
        return (t, e) -> e instanceof Body f
                ? v.apply(t, f)
                : t;
    }

    /**
     * Creates a visiting function for blocks.
     *
     * @param v   the block visitor
     * @param <T> accumulator type
     * @return the visiting function for blocks
     */
    static <T> BiFunction<T, CodeElement<?, ?>, T> blockVisitor(BiFunction<T, Block, T> v) {
        return (t, e) -> e instanceof Block f
                ? v.apply(t, f)
                : t;
    }

    /**
     * Creates a visiting function for operations.
     *
     * @param v   the operation visitor
     * @param <T> accumulator type
     * @return the visiting function for operations
     */
    static <T> BiFunction<T, CodeElement<?, ?>, T> opVisitor(BiFunction<T, Op, T> v) {
        return (t, e) -> e instanceof Op f
                ? v.apply(t, f)
                : t;
    }

    /**
     * Returns the parent code element.
     * <p>
     * If this element is an instance of {@code Op} then the parent may be {@code null}
     * if operation is not assigned to a block.
     *
     * @return the parent code element
     */
    CodeElement<?, E> parent();

    /**
     * Returns the child code elements, as an unmodifiable list.
     *
     * @return the child code elements
     */
    List<C> children();
}
