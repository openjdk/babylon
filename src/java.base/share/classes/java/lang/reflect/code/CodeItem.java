package java.lang.reflect.code;

/**
 * A code item, one of {@link CodeElement}, {@link Value}, or {@link TypeElement}.
 */
public sealed interface CodeItem
        permits CodeElement, Value, TypeElement {
    // @@@ Common functionality between elements and values?
}
