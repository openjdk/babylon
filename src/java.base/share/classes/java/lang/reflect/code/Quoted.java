/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package java.lang.reflect.code;

import java.util.Map;

/**
 * The quoted form of an operation.
 * <p>
 * The quoted form is utilized when the code model of some code is to be obtained rather than obtaining the result of
 * executing that code. For example passing the of a lambda expression in quoted form rather than the expression being
 * targeted to a functional interface from which it can be invoked.
 */
public final class Quoted {
    private final Op op;
    private final Map<Value, Object> capturedValues;

    /**
     * Constructs the quoted form of a given invokable operation.
     *
     * @param op the invokable operation.
     */
    public Quoted(Op op) {
        this(op, Map.of());
    }

    /**
     * Constructs the quoted form of a given invokable operation.
     *
     * @param op             the invokable operation.
     * @param capturedValues the capture values referred to by the operation
     */
    public Quoted(Op op, Map<Value, Object> capturedValues) {
        this.op = op;
        this.capturedValues = Map.copyOf(capturedValues);
    }

    /**
     * Returns the invokable operation.
     *
     * @return the invokable operation.
     */
    public Op op() {
        return op;
    }

    /**
     * Returns the captured values.
     *
     * @return the captured values, as an unmodifiable map.
     */
    public Map<Value, Object> capturedValues() {
        return capturedValues;
    }

    public static Quoted quote(Op t) {
        return new Quoted(t);
    }
}
