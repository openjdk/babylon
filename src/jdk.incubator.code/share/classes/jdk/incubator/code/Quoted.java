/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.  Oracle designates this
 * particular file as subject to the "Classpath" exception as provided
 * by Oracle in the LICENSE file that accompanied this code.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

package jdk.incubator.code;

import java.util.*;

/**
 * The quoted form of an operation.
 * <p>
 * The quoted form is utilized when the code model of some code is to be obtained rather than obtaining the result of
 * executing that code. For example passing the of a lambda expression in quoted form rather than the expression being
 * targeted to a functional interface from which it can be invoked.
 */
public final class Quoted {
    private final Op op;
    private final SequencedMap<Value, Object> capturedValues;

    static final SequencedMap<Value, Object> EMPTY_SEQUENCED_MAP = new LinkedHashMap<>();

    /**
     * Constructs the quoted form of a given operation.
     *
     * @param op the invokable operation.
     */
    public Quoted(Op op) {
        this(op, EMPTY_SEQUENCED_MAP);
    }

    /**
     * Constructs the quoted form of a given operation.
     * <p>
     * The captured values key set must have the same elements and same encounter order as
     * operation's captured values, specifically the following expression should evaluate to true:
     * {@snippet lang=java :
     * op.capturedValues().equals(new ArrayList<>(capturedValues.keySet()));
     * }
     *
     * @param op             the operation.
     * @param capturedValues the captured values referred to by the operation
     * @see Op#capturedValues()
     */
    public Quoted(Op op, SequencedMap<Value, Object> capturedValues) {
        // @@@ This check is potentially expensive, remove or keep as assert?
        // @@@ Or make Quoted an interface, with a module private implementation?
        assert op.capturedValues().equals(new ArrayList<>(capturedValues.keySet()));
        this.op = op;
        this.capturedValues = Collections.unmodifiableSequencedMap(new LinkedHashMap<>(capturedValues));
    }

    /**
     * Returns the operation.
     *
     * @return the operation.
     */
    public Op op() {
        return op;
    }

    /**
     * Returns the captured values.
     * <p>
     * The captured values key set has the same elements and same encounter order as
     * operation's captured values, specifically the following expression evaluates to true:
     * {@snippet lang=java :
     * op().capturedValues().equals(new ArrayList<>(capturedValues().keySet()));
     * }
     *
     * @return the captured values, as an unmodifiable map.
     */
    public SequencedMap<Value, Object> capturedValues() {
        return capturedValues;
    }
}
