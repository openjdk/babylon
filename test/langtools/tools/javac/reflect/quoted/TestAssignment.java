/*
 * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 *
 * This code is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License version 2 only, as
 * published by the Free Software Foundation.
 *
 * This code is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 * version 2 for more details (a copy is included in the LICENSE file that
 * accompanied this code).
 *
 * You should have received a copy of the GNU General Public License version
 * 2 along with this work; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
 * or visit www.oracle.com if you need additional information or have any
 * questions.
 */

/*
 * @test /nodynamiccopyright/
 * @modules jdk.incubator.code
 * @compile/fail/ref=TestAssignment.out -Xlint:-incubating -XDrawDiagnostics TestAssignment.java
 */

import jdk.incubator.code.Quoted;

class TestAssignment {
    void test(boolean cond) {
        Quoted f_NoRet = () -> {};
        Quoted fiS_NoRet = (int i) -> ""; // ok (int->String)
        Quoted fiS_Ret = (int i) -> { return ""; }; // ok (int->String)
        Quoted fiV_NoRet = (int i) -> { }; // ok (int->V)
        Quoted fiV_Ret = (int i) -> { return; }; // ok (int->V)
        Quoted fiV_RetRet = (int i) -> { if (cond) return; else return; }; // ok (int->V)
        Quoted fiS_RetRet = (int i) -> { if (cond) return "1"; else return "2"; }; // ok (int->String)
    }

    void testImplicit(boolean cond) {
        Quoted fiS_NoRet = (i) -> ""; // error - no parameter types
        Quoted fiS_Ret = (i) -> { return ""; }; // error - no parameter types
        Quoted fiV_NoRet = (i) -> { }; // error - no parameter types
        Quoted fiV_Ret = (i) -> { return; }; // error - no parameter types
        Quoted fiV_RetRet = (i) -> { if (cond) return; else return; }; // error - no parameter types
        Quoted fiS_RetRet = (i) -> { if (cond) return "1"; else return "2"; }; // error - no parameter types
    }

    void testImplicitVar(boolean cond) {
        Quoted fiS_NoRet = (var i) -> ""; // error - no parameter types
        Quoted fiS_Ret = (var i) -> { return ""; }; // error - no parameter types
        Quoted fiV_NoRet = (var i) -> { }; // error - no parameter types
        Quoted fiV_Ret = (var i) -> { return; }; // error - no parameter types
        Quoted fiV_RetRet = (var i) -> { if (cond) return; else return; }; // error - no parameter types
        Quoted fiS_RetRet = (var i) -> { if (cond) return "1"; else return "2"; }; // error - no parameter types
    }

    void testBadInferredReturn(boolean cond) {
        Quoted fi_RetVRetS = (int i) -> { if (cond) return; else return ""; }; // error - only one branch returns
        Quoted fi_RetS = (int i) -> { if (cond) { return "2"; } }; // error - one return, but body completes normally
    }
}