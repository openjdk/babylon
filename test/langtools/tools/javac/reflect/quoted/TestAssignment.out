TestAssignment.java:44:28: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:45:26: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:46:28: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:47:26: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:48:29: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:49:29: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:53:28: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:54:26: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:55:28: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:56:26: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:57:29: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:58:29: compiler.err.quoted.lambda.must.be.explicit
TestAssignment.java:62:41: compiler.err.cant.infer.quoted.lambda.return.type: void,java.lang.String
TestAssignment.java:63:37: compiler.err.cant.infer.quoted.lambda.return.type: java.lang.String,void
14 errors
